rolling-shutter/collator/batch/batchhandler.go

```diff
@@ -0,0 +1,503 @@
+// The function also depends on the current mainchain block number and the configured execution
+// block delay. The result will encode a block number and a sequence number. The sequence number
+// will be the sequence number of the previous epoch id plus one. The block number will be
+// max(current block number - execution block delay, block number encoded in previous epoch id, 0).
+func computeNextEpochID(epochID uint64, currentBlockNumber uint32, executionBlockDelay uint32) uint64 {
+	executionBlockNumber := uint32(0)
+	if currentBlockNumber >= executionBlockDelay {
+		executionBlockNumber = currentBlockNumber - executionBlockDelay
+	}
+
+	previousExecutionBlockNumber := epochid.BlockNumber(epochID)
+	if executionBlockNumber < previousExecutionBlockNumber {
+		executionBlockNumber = previousExecutionBlockNumber
+	}
+
+	sequenceNumber := epochid.SequenceNumber(epochID)
+	return epochid.New(sequenceNumber+1, executionBlockNumber)
+}
+
+// GetNextEpochID gets the epochID that will be used for the next decryption trigger or cipher batch.
+func GetNextEpochID(ctx context.Context, db *cltrdb.Queries) (uint64, error) {
+	epochID, err := db.GetNextEpochID(ctx)
+	if err != nil {
+		// There should already be an epochID in the database so not finding a row is an error
+		return 0, err
+	}
+	return shdb.DecodeUint64(epochID), nil
+}
+
+func NewBatchHandler(cfg config.Config, dbpool *pgxpool.Pool) (*BatchHandler, error) {
+	ctx := context.Background()
+
+	l2Client, err := rpc.Dial(cfg.SequencerURL)
+	if err != nil {
+		return nil, err
+	}
+	l2EthClient := ethclient.NewClient(l2Client)
+	chainID, err := l2EthClient.ChainID(ctx)
+	if err != nil {
+		return nil, err
+	}
+	signer := txtypes.LatestSignerForChainID(chainID)
+
+	return &BatchHandler{
+		l2Client:    l2Client,
+		l2EthClient: l2EthClient,
+		config:      cfg,
+		signer:      signer,
+		txpool:      NewTransactionPool(signer),
+		dbpool:      dbpool,
+	}, nil
+}
+
+// BatchHandler is a threadsafe handler to process the following actions
+// - validate and enqueue incoming tx's either to the currently pending batch
+//    or for later validation in the transaction pool
+// - handle incoming decryption keys and send the currently pending batch-transaction
+//    to the sequencer
+// - start next epoch and initiate the broadcasting of the decryption-trigger to
+//    the keypers
+type BatchHandler struct {
+	mux      sync.Mutex
+	l2Client *rpc.Client
+	// l2EthClient is the RPC l2Client wrapped by ethclient.Client
+	l2EthClient *ethclient.Client
+	config      config.Config
+	txpool      *TransactionPool
+	signer      txtypes.Signer
+	dbpool      *pgxpool.Pool
+	latestBatch *Batch
+}
+
+func (bh *BatchHandler) Signer() txtypes.Signer {
+	return bh.signer
+}
+
+// LatestEpochID returns the epoch-id associated to
+// the current batch that is yet to be submitted to the sequencer.
+func (bh *BatchHandler) LatestEpochID() uint64 {
+	return bh.latestBatch.EpochID()
```

Comment:
It is set by `bh.InitialiseEpoch()` which is expected to be called on shutter node initialisation.
The reason for having to call this explicitly is because initialising the batch includes querying the sequencer to get the current nonce state etc.
This might deserve a comment though